More optimizations on metal

ikawrakow opened the pull request:
On 30-core M2 Max:

| model                        | backend  | test     |    t/s (Master)|     t/s (PR)   |   Speedup|
|------------------------------|----------|----------|---------------:|---------------:|---------:|
| LLaMA 7B mostly Q4_0         | Metal    | tg 32    |    62.16 ± 0.07| 62.57 ± 0.08   |   1.007  |
| LLaMA 7B mostly Q4_0         | Metal    | tg 64    |    61.65 ± 0.05| 62.09 ± 0.08   |   1.007  |
| LLaMA 7B mostly Q4_0         | Metal    | tg 128   |    61.22 ± 0.13| 61.71 ± 0.05   |   1.008  |
| LLaMA 7B mostly Q4_0         | Metal    | tg 256   |    58.44 ± 0.03| 59.46 ± 0.15   |   1.017  |
| LLaMA 7B mostly Q4_0         | Metal    | pp 32    |   382.65 ± 1.92| 388.64 ± 1.47  |   1.016  |
| LLaMA 7B mostly Q4_0         | Metal    | pp 64    |   450.66 ± 1.26| 468.87 ± 1.91  |   1.040  |
| LLaMA 7B mostly Q4_0         | Metal    | pp 128   |   444.64 ± 0.62| 484.74 ± 0.49  |   1.090  |
| LLaMA 7B mostly Q4_0         | Metal    | pp 256   |   406.50 ± 0.21| 479.20 ± 0.55  |   1.179  |
| LLaMA 7B mostly Q4_0         | Metal    | pp 512   |   327.79 ± 0.18| 433.35 ± 0.17  |   1.322  |
| LLaMA 7B mostly Q4_0         | Metal    | pp 1024  |   227.97 ± 0.10| 352.67 ± 0.04  |   1.547  |

With these changes along with the merged #2951, perplexity now runs in 13.6 minutes on my M2 Max laptop vs ~24 minutes before.

Commit:
Very minor speedup via simd-group synchronization in f16 x f32

Commit:
Another very minor speedup on metal

Commit:
Quite significant PP speedup on metal

Commit:
Another attempt

Commit:
Minor

ggerganov committed:
Merge branch 'master' into ik/more_metal_optimizations

ggerganov commented:
M2 Ultra results:

| model         |       size | test       |    master    t/s |    IK        t/s | speedup   |
| --------------| ---------: | ---------- | ---------------: | ---------------: | --------: |
| LLaMA 7B Q4_0 |   3.56 GiB | pp 32      |    701.75 ± 4.36 |    712.61 ± 2.79 |    1.015  |
| LLaMA 7B Q4_0 |   3.56 GiB | pp 64      |    879.19 ± 3.16 |    919.36 ± 5.79 |    1.046  |
| LLaMA 7B Q4_0 |   3.56 GiB | pp 128     |    964.09 ± 1.39 |   1060.88 ± 1.97 |    1.100  |
| LLaMA 7B Q4_0 |   3.56 GiB | pp 256     |    888.98 ± 1.02 |   1073.01 ± 1.69 |    1.207  |
| LLaMA 7B Q4_0 |   3.56 GiB | pp 512     |    719.63 ± 0.23 |   1009.82 ± 0.96 |    1.403  |
| LLaMA 7B Q4_0 |   3.56 GiB | pp 1024    |    549.62 ± 0.86 |    844.49 ± 1.62 |    1.536  |
| LLaMA 7B Q4_0 |   3.56 GiB | tg 128     |     87.45 ± 0.10 |     87.65 ± 0.17 |    1.002  |

Commit:
Massive improvement for TG for fp16

ikawrakow commented:
With the latest commit https://github.com/ggerganov/llama.cpp/pull/2959/commits/363f0bf5580d9d59aaabfb227f6595f4806b32ab TG for fp16 is basically 2X faster compared to master. PP is also improved by a margin rapidly increasing with context length.

On 30-core M2-Max:

| model                 | backend    | test       |    t/s (Master)  |   t/s (PR)   |   Speedup |
| --------------------- | ---------- | ---------- | ---------------: | -----------: | --------: |
| LLaMA 7B mostly F16   | Metal      | tg 32      |     12.04 ± 0.01 | 24.34 ± 0.05 |  2.022    |
| LLaMA 7B mostly F16   | Metal      | tg 64      |     12.01 ± 0.00 | 24.28 ± 0.03 |  2.022    |
| LLaMA 7B mostly F16   | Metal      | tg 128     |     11.91 ± 0.09 | 24.05 ± 0.11 |  2.020    |
| LLaMA 7B mostly F16   | Metal      | tg 256     |     10.88 ± 0.91 | 23.83 ± 0.11 |  2.190    |
| LLaMA 7B mostly F16   | Metal      | pp 32      |    426.51 ± 2.98 | 433.79 ± 2.54|  1.017    |
| LLaMA 7B mostly F16   | Metal      | pp 64      |    509.79 ± 1.16 | 531.98 ± 1.61|  1.044    |
| LLaMA 7B mostly F16   | Metal      | pp 128     |    503.01 ± 0.99 | 548.36 ± 0.98|  1.090    |
| LLaMA 7B mostly F16   | Metal      | pp 256     |    449.66 ± 0.48 | 543.37 ± 0.49|  1.208    |
| LLaMA 7B mostly F16   | Metal      | pp 512     |    353.80 ± 0.10 | 494.99 ± 0.42|  1.399    |
| LLaMA 7B mostly F16   | Metal      | pp 1024    |    266.73 ± 0.09 | 424.11 ± 0.18|  1.590    |

@ggerganov I'm curious to know how this compares to  #2891 on you M2 Ultra.

ggerganov commented:
| model                   |       size | test       |     #2891    t/s |     PR       t/s | speedup   |
| ----------------------- | ---------: | ---------- | ---------------: | ---------------: | --------: |
| LLaMA 7B F16            |  12.55 GiB | pp 32      |    717.82 ± 4.71 |    736.82 ± 2.22 |    1.026  |
| LLaMA 7B F16            |  12.55 GiB | pp 64      |    974.19 ± 3.94 |   1040.40 ± 5.28 |    1.068  |
| LLaMA 7B F16            |  12.55 GiB | pp 128     |   1049.52 ± 1.58 |   1193.71 ± 2.24 |    1.137  |
| LLaMA 7B F16            |  12.55 GiB | pp 256     |    960.61 ± 1.25 |   1214.15 ± 0.90 |    1.264  |
| LLaMA 7B F16            |  12.55 GiB | pp 512     |    763.64 ± 0.41 |   1135.89 ± 0.25 |    1.487  |
| LLaMA 7B F16            |  12.55 GiB | pp 1024    |    574.10 ± 1.05 |    965.57 ± 0.21 |    1.682  |
| LLaMA 7B F16            |  12.55 GiB | tg 32      |     41.57 ± 0.02 |     40.69 ± 0.04 |    0.979  |
| LLaMA 7B F16            |  12.55 GiB | tg 64      |     41.51 ± 0.03 |     40.58 ± 0.03 |    0.978  |
| LLaMA 7B F16            |  12.55 GiB | tg 128     |     41.25 ± 0.03 |     40.27 ± 0.04 |    0.976  |
| LLaMA 7B F16            |  12.55 GiB | tg 256     |     40.88 ± 0.01 |     39.89 ± 0.08 |    0.976  |

Q4_0 7B Perplexity time also dropped - is this expected?

```
perplexity: calculating perplexity over 655 chunks, batch_size=512
perplexity: 0.64 seconds per pass - ETA 6.97 minutes
```

Also, if F16 and Q4_0 `pp` speed is comparable (F16 even faster now), how come the F16 perplexity is so much slower?

```
perplexity: calculating perplexity over 655 chunks, batch_size=512
perplexity: 0.85 seconds per pass - ETA 9.25 minutes
```

Is this a wrong ETA calculation, or do we have some significant overhead somewhere?

ikawrakow commented:
> Also, if F16 and Q4_0 pp speed is comparable (F16 even faster now), how come the F16 perplexity is so much slower?

My experience with the ETA is that it is not very accurate. For `Q4_0` it predicted 14.5 minutes but then finished in 13.6 on my M2 Max. For `fp16` it is predicting 15.5 minutes, and I was also wondering why it is slower than `Q4_0` given that t/s for `fp16` is not better than `Q4_0`. I guess, I'll let it run to completion.

ggerganov commented:
I'm also running.

Q4_0 finished in just 5m and 28s:

<details>

```
system_info: n_threads = 16 / 24 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | SSSE3 = 0 | VSX = 0 | 
perplexity: tokenizing the input ..
perplexity: tokenization took 555.092 ms
perplexity: calculating perplexity over 655 chunks, batch_size=512
perplexity: 0.63 seconds per pass - ETA 6.90 minutes
[1]4.4083,[2]4.9137,[3]5.8087,[4]6.4334,[5]6.5183,[6]6.4580,[7]6.6552,[8]6.7720,[9]7.1091,[10]7.3490,[11]7.5538,[12]7.5747,[13]7.5065,[14]7.5835,[15]7.8312,[16]7.4352,[17]7.3219,[18]7.2706,[19]6.9050,[20]6.8973,[21]6.7935,[22]6.6206,[23]6.5822,[24]6.4818,[25]6.4785,[26]6.3148,[27]6.1317,[28]6.0291,[29]5.9307,[30]5.7704,[31]5.7397,[32]5.7649,[33]5.7076,[34]5.7423,[35]5.7684,[36]5.8032,[37]5.8046,[38]5.8130,[39]5.8435,[40]5.9025,[41]5.9135,[42]5.9526,[43]5.9132,[44]5.9670,[45]5.9734,[46]5.9428,[47]5.9681,[48]5.9435,[49]5.9497,[50]5.9054,[51]5.9008,[52]5.8909,[53]5.9358,[54]5.9171,[55]5.8918,[56]5.9238,[57]5.9458,[58]5.9670,[59]5.9849,[60]6.0304,[61]6.0207,[62]6.0783,[63]6.1122,[64]6.1242,[65]6.1716,[66]6.1813,[67]6.1964,[68]6.2132,[69]6.2359,[70]6.2670,[71]6.2845,[72]6.3174,[73]6.3773,[74]6.3830,[75]6.3972,[76]6.4099,[77]6.4203,[78]6.4055,[79]6.4338,[80]6.4266,[81]6.4372,[82]6.4417,[83]6.3908,[84]6.3768,[85]6.3654,[86]6.3429,[87]6.2834,[88]6.2565,[89]6.2375,[90]6.2222,[91]6.2463,[92]6.2405,[93]6.2370,[94]6.2343,[95]6.2630,[96]6.2609,[97]6.2558,[98]6.2495,[99]6.2344,[100]6.2330,[101]6.2570,[102]6.2513,[103]6.2718,[104]6.2783,[105]6.2786,[106]6.2934,[107]6.2914,[108]6.3017,[109]6.2958,[110]6.2935,[111]6.3143,[112]6.3351,[113]6.3345,[114]6.3311,[115]6.3370,[116]6.3280,[117]6.3325,[118]6.3605,[119]6.3806,[120]6.4151,[121]6.4306,[122]6.4553,[123]6.4922,[124]6.5114,[125]6.5023,[126]6.5425,[127]6.5792,[128]6.6103,[129]6.5937,[130]6.6009,[131]6.5956,[132]6.5870,[133]6.5722,[134]6.5813,[135]6.5770,[136]6.5644,[137]6.5573,[138]6.5403,[139]6.5300,[140]6.5262,[141]6.4979,[142]6.4947,[143]6.4670,[144]6.4459,[145]6.4373,[146]6.4251,[147]6.4296,[148]6.4298,[149]6.4246,[150]6.4211,[151]6.4223,[152]6.4120,[153]6.3948,[154]6.3864,[155]6.3928,[156]6.3884,[157]6.4047,[158]6.4087,[159]6.4146,[160]6.4172,[161]6.4286,[162]6.3996,[163]6.3869,[164]6.3628,[165]6.3322,[166]6.3052,[167]6.2681,[168]6.2369,[169]6.2231,[170]6.2126,[171]6.1857,[172]6.1688,[173]6.1513,[174]6.1222,[175]6.0995,[176]6.0869,[177]6.0665,[178]6.0437,[179]6.0267,[180]6.0175,[181]5.9964,[182]5.9792,[183]5.9655,[184]5.9647,[185]5.9573,[186]5.9577,[187]5.9633,[188]5.9594,[189]5.9768,[190]5.9788,[191]6.0009,[192]6.0171,[193]6.0334,[194]6.0449,[195]6.0670,[196]6.0826,[197]6.1057,[198]6.1203,[199]6.1233,[200]6.1287,[201]6.1237,[202]6.1438,[203]6.1513,[204]6.1506,[205]6.1611,[206]6.1686,[207]6.1646,[208]6.1732,[209]6.1778,[210]6.1837,[211]6.1940,[212]6.2008,[213]6.2113,[214]6.2143,[215]6.2176,[216]6.2309,[217]6.2486,[218]6.2621,[219]6.2624,[220]6.2584,[221]6.2533,[222]6.2501,[223]6.2409,[224]6.2337,[225]6.2299,[226]6.2499,[227]6.2586,[228]6.2635,[229]6.2683,[230]6.2653,[231]6.2817,[232]6.2692,[233]6.2521,[234]6.2380,[235]6.2225,[236]6.2162,[237]6.2059,[238]6.2084,[239]6.1932,[240]6.1836,[241]6.1861,[242]6.1902,[243]6.1890,[244]6.1771,[245]6.1734,[246]6.1622,[247]6.1499,[248]6.1419,[249]6.1391,[250]6.1433,[251]6.1355,[252]6.1311,[253]6.1216,[254]6.1171,[255]6.1056,[256]6.0882,[257]6.0756,[258]6.0671,[259]6.0648,[260]6.0565,[261]6.0521,[262]6.0471,[263]6.0413,[264]6.0232,[265]6.0226,[266]6.0199,[267]6.0134,[268]6.0228,[269]6.0207,[270]6.0217,[271]6.0295,[272]6.0338,[273]6.0345,[274]6.0362,[275]6.0448,[276]6.0507,[277]6.0659,[278]6.0760,[279]6.0856,[280]6.0887,[281]6.0982,[282]6.1048,[283]6.1190,[284]6.1267,[285]6.1351,[286]6.1488,[287]6.1477,[288]6.1537,[289]6.1450,[290]6.1292,[291]6.1140,[292]6.0991,[293]6.0858,[294]6.0878,[295]6.0870,[296]6.0914,[297]6.0905,[298]6.0937,[299]6.0909,[300]6.0803,[301]6.0806,[302]6.0729,[303]6.0645,[304]6.0564,[305]6.0531,[306]6.0411,[307]6.0434,[308]6.0463,[309]6.0303,[310]6.0248,[311]6.0186,[312]6.0212,[313]6.0154,[314]6.0133,[315]5.9974,[316]5.9929,[317]5.9771,[318]5.9569,[319]5.9690,[320]5.9811,[321]5.9855,[322]5.9809,[323]5.9744,[324]5.9714,[325]5.9816,[326]5.9814,[327]5.9836,[328]5.9875,[329]5.9927,[330]5.9954,[331]6.0080,[332]6.0053,[333]6.0121,[334]6.0065,[335]6.0005,[336]6.0037,[337]6.0011,[338]6.0009,[339]5.9953,[340]5.9917,[341]5.9987,[342]6.0016,[343]6.0073,[344]6.0074,[345]6.0076,[346]6.0049,[347]6.0097,[348]6.0127,[349]6.0150,[350]6.0125,[351]6.0135,[352]6.0136,[353]6.0082,[354]6.0083,[355]6.0133,[356]6.0164,[357]6.0127,[358]6.0216,[359]6.0249,[360]6.0211,[361]6.0208,[362]6.0277,[363]6.0391,[364]6.0451,[365]6.0501,[366]6.0515,[367]6.0600,[368]6.0574,[369]6.0582,[370]6.0598,[371]6.0550,[372]6.0598,[373]6.0643,[374]6.0629,[375]6.0630,[376]6.0696,[377]6.0660,[378]6.0688,[379]6.0749,[380]6.0670,[381]6.0635,[382]6.0585,[383]6.0573,[384]6.0569,[385]6.0559,[386]6.0550,[387]6.0549,[388]6.0515,[389]6.0462,[390]6.0390,[391]6.0316,[392]6.0280,[393]6.0260,[394]6.0291,[395]6.0276,[396]6.0204,[397]6.0276,[398]6.0308,[399]6.0385,[400]6.0377,[401]6.0392,[402]6.0401,[403]6.0419,[404]6.0484,[405]6.0384,[406]6.0344,[407]6.0338,[408]6.0353,[409]6.0474,[410]6.0583,[411]6.0693,[412]6.0854,[413]6.0963,[414]6.1037,[415]6.1091,[416]6.1169,[417]6.1286,[418]6.1324,[419]6.1390,[420]6.1473,[421]6.1584,[422]6.1625,[423]6.1694,[424]6.1802,[425]6.1886,[426]6.1952,[427]6.1996,[428]6.2083,[429]6.2127,[430]6.2215,[431]6.2351,[432]6.2395,[433]6.2389,[434]6.2339,[435]6.2350,[436]6.2374,[437]6.2471,[438]6.2545,[439]6.2518,[440]6.2511,[441]6.2462,[442]6.2448,[443]6.2457,[444]6.2459,[445]6.2435,[446]6.2459,[447]6.2483,[448]6.2525,[449]6.2499,[450]6.2509,[451]6.2470,[452]6.2355,[453]6.2272,[454]6.2216,[455]6.2227,[456]6.2272,[457]6.2293,[458]6.2269,[459]6.2271,[460]6.2354,[461]6.2328,[462]6.2316,[463]6.2356,[464]6.2345,[465]6.2316,[466]6.2241,[467]6.2242,[468]6.2243,[469]6.2262,[470]6.2267,[471]6.2220,[472]6.2261,[473]6.2206,[474]6.2220,[475]6.2164,[476]6.2181,[477]6.2109,[478]6.2098,[479]6.2166,[480]6.2212,[481]6.2230,[482]6.2186,[483]6.2145,[484]6.2167,[485]6.2148,[486]6.2096,[487]6.2095,[488]6.2071,[489]6.2024,[490]6.2002,[491]6.1974,[492]6.1914,[493]6.1886,[494]6.1870,[495]6.1872,[496]6.1836,[497]6.1781,[498]6.1764,[499]6.1721,[500]6.1628,[501]6.1562,[502]6.1568,[503]6.1558,[504]6.1468,[505]6.1494,[506]6.1505,[507]6.1450,[508]6.1406,[509]6.1398,[510]6.1431,[511]6.1475,[512]6.1511,[513]6.1532,[514]6.1596,[515]6.1543,[516]6.1533,[517]6.1547,[518]6.1547,[519]6.1581,[520]6.1601,[521]6.1618,[522]6.1647,[523]6.1654,[524]6.1711,[525]6.1745,[526]6.1757,[527]6.1775,[528]6.1726,[529]6.1731,[530]6.1680,[531]6.1666,[532]6.1714,[533]6.1738,[534]6.1724,[535]6.1747,[536]6.1693,[537]6.1669,[538]6.1716,[539]6.1726,[540]6.1762,[541]6.1767,[542]6.1780,[543]6.1794,[544]6.1806,[545]6.1783,[546]6.1789,[547]6.1747,[548]6.1695,[549]6.1696,[550]6.1669,[551]6.1633,[552]6.1612,[553]6.1572,[554]6.1552,[555]6.1524,[556]6.1521,[557]6.1544,[558]6.1505,[559]6.1496,[560]6.1494,[561]6.1495,[562]6.1469,[563]6.1470,[564]6.1508,[565]6.1524,[566]6.1521,[567]6.1502,[568]6.1507,[569]6.1493,[570]6.1520,[571]6.1523,[572]6.1532,[573]6.1531,[574]6.1492,[575]6.1488,[576]6.1490,[577]6.1474,[578]6.1457,[579]6.1461,[580]6.1393,[581]6.1356,[582]6.1343,[583]6.1351,[584]6.1356,[585]6.1282,[586]6.1215,[587]6.1222,[588]6.1271,[589]6.1325,[590]6.1352,[591]6.1373,[592]6.1359,[593]6.1329,[594]6.1339,[595]6.1318,[596]6.1356,[597]6.1334,[598]6.1308,[599]6.1328,[600]6.1326,[601]6.1313,[602]6.1326,[603]6.1355,[604]6.1364,[605]6.1398,[606]6.1415,[607]6.1402,[608]6.1367,[609]6.1371,[610]6.1405,[611]6.1384,[612]6.1412,[613]6.1374,[614]6.1325,[615]6.1254,[616]6.1280,[617]6.1218,[618]6.1166,[619]6.1112,[620]6.0975,[621]6.0908,[622]6.0892,[623]6.0910,[624]6.0918,[625]6.0917,[626]6.0907,[627]6.0927,[628]6.0931,[629]6.0926,[630]6.0956,[631]6.1013,[632]6.1065,[633]6.1050,[634]6.1085,[635]6.1095,[636]6.1067,[637]6.1034,[638]6.1059,[639]6.1029,[640]6.1041,[641]6.1045,[642]6.1111,[643]6.1131,[644]6.1142,[645]6.1125,[646]6.1168,[647]6.1135,[648]6.1143,[649]6.1143,[650]6.1182,[651]6.1234,[652]6.1245,[653]6.1284,[654]6.1219,[655]6.1213,
Final estimate: PPL = 6.1213 +/- 0.03511

llama_print_timings:        load time =  1930.79 ms
llama_print_timings:      sample time =     0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings: prompt eval time = 323972.82 ms / 335360 tokens (    0.97 ms per token,  1035.15 tokens per second)
llama_print_timings:        eval time =     0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time = 327603.66 ms
ggml_metal_free: deallocating

real	5m27.892s
user	0m29.329s
sys	0m2.167s
```
</details>

Running F16 ...

ikawrakow commented:
@ggerganov 

I see TG is slightly faster with #2891 on your M2 Ultra. On my M2 Max this PR is very slightly faster (~1%) than #2891 for LLaMA-7B and fior Falcon-7B.

ggerganov commented:
> I see TG is slightly faster with https://github.com/ggerganov/llama.cpp/pull/2891 on your M2 Ultra. On my M2 Max this PR is very slightly faster (~1%) than https://github.com/ggerganov/llama.cpp/pull/2891 for LLaMA-7B and fior Falcon-7B.

Have you fetched the latest branch of #2891 ? I think I updated it yesterday by merging `master` into it and if you haven't fetched since then, then you might have a stale version. Or it's just a small difference between Max and Ultra

ikawrakow commented:
> > I see TG is slightly faster with #2891 on your M2 Ultra. On my M2 Max this PR is very slightly faster (~1%) than #2891 for LLaMA-7B and fior Falcon-7B.
> 
> Have you fetched the latest branch of #2891 ? I think I updated it yesterday by merging `master` into it and if you haven't fetched since then, then you might have a stale version. Or it's just a small difference between Max and Ultra

Yes, I just updated this afternoon. I'm on b46ae7b for #2891

Commit:
~4-5% improvement for Q8_0 TG on metal

ggerganov commented on ggml-metal.metal:
```diff
@@ -528,42 +535,102 @@ kernel void kernel_mul_mat_f16_f32(
-    uint ith = tpitg.x;
-    uint nth = tptg.x;
+    float sumf = 0;
+    if (ne00 < 128) {
```
Should we have 2 separate kernels to avoid this branch?

ikawrakow replied:
I was considering it, but taking into account the desire for shorter and simple code didn't do it. This is something one needs to study more carefully anyway. The best way to perform this computation is not just a function of `ne00`.

mechanicmuthu commented:
Causing incoherent paragraphs very fast in M2 Metal inference. -ngl 0 works fine.

Sample: (With -ngl 99)
[INST]
<<SYS>>
You are a helpful assistant.
<</SYS>>
Tenali Rama story. 
 [/INST]
Certainly, I'd be happy to tell you a story about Tenali Rama!

Tenali Rama was a great poet and storyteller from India who lived in the 16th century. He was known for his wit and humor, and was particularly famous for his ability to create humorous stories and poems that poked fun at the foolishness of King and aristocrats of his time.
One day,One day, Rama was at the king's court of the king Krishnadev, who was known for a s good and foolish king. He loved to beed to and commerce and of his subjects,  and did and making foolish all the time. Rama was not only clever than king's foolish his, but also very poor.  and he had. He was very and 64 and king's and 4 queens, and  and that's why Rama 6 for . sake. The king   he is want Rama, so  He was . Not ) i, he befoolish  And  he the king  Rama)  for him a very kind of do.  him. 

 him  the king, and to them  of place, they still On themeant their poor  have were  They the to eat there  and drinking, smile  and the  and and the them the for).



The king Rama Rama 6 him a  the  to the  his  and the king with  a  The the   his   and

Rama Rama six years king Krish his his in the was the Rama is was a the not  his to poor Rama. He had his the  a  king of and he the king He  was  no  foolish king and  four king  and the or for   and the  king of Ram at all the king Wash   in  king Rama   for king Hindustan so and king all   Rama   and   all  a king all  for a  a king    Rama  a king for  day.


 kingish Rama 6 Rama Rama   Of   Raman Rams years  por the king  years. Rams poor king Rams  and he Rams years yearsRams, of poor Rams six years were   years, the years   king king Rams.
 
 
 
 Sample with -ngl 0:
 
 
 [INST]
<<SYS>>
You are a helpful assistant.
<</SYS>>
Tenali Rama story. 
 [/INST]
Ah, the delightful Tenali Rama! He was a renowned poet and humorist of the 18th century in India. Here's a brief summary of his life and works:
Tenali Rama (1725-1805) was born in a brahmin family in the state of Andhra Pradesh, India. He was largely self-taught and developed a passion for literature at an early age. He was known for his wit, humor, and creativity, which he showcased through his writings, including poetry, drama, and fiction.
Rama's works are characterized by their playfulness, clever wordplay, and satire. He often used humor to comment on contemporary issues, politics, and society. His most notable work is the "Tenali Ramakrishnudu" (1775), a collection of humorous anecdotes and stories that are still popular today.
Some of Rama's most famous works include:
1. "Tenali Ramakrishnudu" (1775) - This is Rama's magnum opus, a collection of humorous anecdotes and stories that are still popular today. The work is divided into four volumes, each containing a series of tales featuring Tenali Ramakrishnudu, a clever and witty Brahmin who often outsmarts his neighbors and friends with his quick wit and clever ideas.

..

Please revert.

ggerganov commented:
Yes, I just observed that F16 inference is broken with the PR - starts ok and then degrades into incoherent text.
Reverting the commit fixes the issue.

I'm looking into it

ggerganov commented on ggml-metal.metal:
After this commit, text generation becomes incoherent

ggerganov commented on ggml-metal.metal:
```diff
@@ -528,42 +535,102 @@ kernel void kernel_mul_mat_f16_f32(
+
+            float all_sum = simd_sum(sumf);
+            if (tiisg == 0) {
+                for (int i = 4*(ne00/4); i < ne00; ++i) sumf += (float) x[i] * y[i];
```
Changing this to `all_sum += (float) x[i] * y[i];` in both branches seems to resolve the issue
